pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cnsfilevolumeclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internal/cnsoperator/cnsfilevolumeclient/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// FileVolumeClients maintains a client to the API
+// server for operations on CnsFileVolumeClient instance.
+// An in-memory cache is used to read instance data.
+// It also contains a per instance lock to handle
+// concurrent operations.
+type FileVolumeClients struct {
+	client client.Client
+	store  cache.Store
+	// Per volume lock for concurrent access to CnsFileVolumeClient instances.
+	// Keys are strings representing volume handles (or SV-PVC names).
+	// Values are individual sync.Mutex locks that need to be held
+	// to make updates to the CnsFileVolumeClient instance on the API server.
+	fileVolumeClientsLock *sync.Map
+}
+
+// InitFileVolumeClients initializes and returns a variable of type FileVolumeClients.
+func InitFileVolumeClients(ctx context.Context) (*FileVolumeClients, error) {
+	log := logger.GetLogger(ctx)
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Err: %v", err)
+		return nil, err
+	}
+	k8sclient, err := k8s.NewClientForGroup(ctx, config, v1alpha1.GroupVersion.Group)
+	if err != nil {
+		log.Errorf("failed to create k8s client. Err: %v", err)
+		return nil, err
+	}
+	informer, err := k8s.GetDynamicInformer(ctx, v1alpha1.GroupVersion.Group, v1alpha1.GroupVersion.Version, "cnsfilevolumeclients", "")
```

Comment:
This is creating a new dynamicSharedInformerFactory and a new informer for `cnsfilevolumeclients`.  Can we use singleton pattern that creates a single dynamicSharedInformerFactory and all the informers are created from that factory instanced? This way we can better manage the informers.

Reply:
Yup makes sense! Made some changes to `pkg/kubernetes/kubernetes.go` to make it a singleton